### PR TITLE
Fix heartbeat races and reset ticker only if configuration changed

### PR DESCRIPTION
**Pull request by gabriel-farache**
This PR fixes the data races detected by @jakub-dzon and it implements a change in the Update behaviour: the ticker will only be reseted if the configuration (more precisely, the PeriodSeconds field) changed.

Signed-off-by: Gabriel Farache [gfarache@redhat.com](mailto:gfarache@redhat.com)

### CODE REVIEW

**Review by pkliczewski — COMMENTED** [review]
/lgtm

File: internal/heartbeat/heartbeat.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ type HeartbeatData struct {
 	hardware                    hw.Hardware
 	osInfo                      *os2.OS
 	previousMutableHardwareInfo *models.HardwareInfo
+	mu                          sync.Mutex
```
[the remark this answers is not in the document]
Comment (gabriel-farache) [reply internal/heartbeat/heartbeat.go:40]: right, I missed this one

**Review by jakub-dzon — APPROVED** [review]

**Review by machacekondra — APPROVED** [review]

**Comment by ydayagi** [comment]
@gabriel-farache I have a few comments about this PR. Hope it will help.

1. I think that the lock inside initTicker is not needed. You already lock in Update.
2. I would call the lock 'configLock' or 'updateLock'. Because it actually locks the whole update and can be used for any item we want to update while in the lock. In general, I also prefer this way - one lock for the whole update.
3. I would put the top in initTicker. All init/update of ticker in one function.
4. Perhaps you can try using this change for fixing the goroutine leak.

**Comment by gabriel-farache** [comment]
> @gabriel-farache I have a few comments about this PR. Hope it will help.
> 
> 1. I think that the lock inside initTicker is not needed. You already lock in Update.
> 2. I would call the lock 'configLock' or 'updateLock'. Because it actually locks the whole update and can be used for any item we want to update while in the lock. In general, I also prefer this way - one lock for the whole update.
> 3. I would put the top in initTicker. All init/update of ticker in one function.
> 4. Perhaps you can try using this change for fixing the goroutine leak.

Thanks for the feedback.
About your 1st point, `initTicker` is also called by the `Start` method. I guess that it's not possible to have simultaneously an `Update` and a `Start` call, so yes, it should be safe to remove the lock inside the `initTiker`
For the 2nd point, what I wanted to actually protect was `s.previousPeriodSeconds`, not the ticker itself so `configLock seems more appropriated indeed
For the 3rd point, I'm not sure to understand, you would split the `initTicker` in 2 parts: 1 for the initialization of the parameters and 1 with the goroutine creation?
For the 4th point, as the PR is merged, I'll do that in a separate ticket :)

**Comment by gabriel-farache** [comment]
> > @gabriel-farache I have a few comments about this PR. Hope it will help.
> > 
> > 1. I think that the lock inside initTicker is not needed. You already lock in Update.
> > 2. I would call the lock 'configLock' or 'updateLock'. Because it actually locks the whole update and can be used for any item we want to update while in the lock. In general, I also prefer this way - one lock for the whole update.
> > 3. I would put the top in initTicker. All init/update of ticker in one function.
> > 4. Perhaps you can try using this change for fixing the goroutine leak.
> 
> Thanks for the feedback @ydayagi About your 1st point, `initTicker` is also called by the `Start` method. I guess that it's not possible to have simultaneously an `Update` and a `Start` call, so yes, it should be safe to remove the lock inside the `initTiker` For the 2nd point, what I wanted to actually protect was `s.previousPeriodSeconds`, not the ticker itself so `configLock` seems more appropriated indeed For the 3rd point, I'm not sure to understand, you would split the `initTicker` in 2 parts: 1 for the initialization of the parameters and 1 with the goroutine creation? For the 4th point, as the PR is merged, I'll do that in a separate ticket :)

About the 3rd point, with the change using RWMutex, it's not the same type of lock that is used in the 2 methods: in `Update` it's a read lock while it's a write lock in `initTicker` so I'll keep both lock

**Comment by ydayagi** [comment]
> > @gabriel-farache I have a few comments about this PR. Hope it will help.
> > 
> > 1. I think that the lock inside initTicker is not needed. You already lock in Update.
> > 2. I would call the lock 'configLock' or 'updateLock'. Because it actually locks the whole update and can be used for any item we want to update while in the lock. In general, I also prefer this way - one lock for the whole update.
> > 3. I would put the top in initTicker. All init/update of ticker in one function.
> > 4. Perhaps you can try using this change for fixing the goroutine leak.
> 
> Thanks for the feedback @ydayagi About your 1st point, `initTicker` is also called by the `Start` method. I guess that it's not possible to have simultaneously an `Update` and a `Start` call, so yes, it should be safe to remove the lock inside the `initTiker` For the 2nd point, what I wanted to actually protect was `s.previousPeriodSeconds`, not the ticker itself so `configLock` seems more appropriated indeed For the 3rd point, I'm not sure to understand, you would split the `initTicker` in 2 parts: 1 for the initialization of the parameters and 1 with the goroutine creation? For the 4th point, as the PR is merged, I'll do that in a separate ticket :)

I meant that the call to 'ticker.Stop', I would put it inside initTicker. Think of it as 2 parts: one is destroying/stopping existing ticker and the second for creating a new one

**Pull request merged**